### DISCUSSION TITLE
Why is the btree's size_type signed?

### DISCUSSION THREAD

**Discussion by bobaoCrbs** [abseil/abseil-cpp#993] [Q&A] [answered]
Hi,
I am just wondering why the size() of btree containers is of type "signed size_t". (instead of just size_t?).
i.e. [this line](https://github.com/abseil/abseil-cpp/blob/5de90e2673bd0c19de67c68e2fe543444dc1114a/absl/container/internal/btree.h#L241)

Thanks
Bo

**Comment by tituswinters** [comment, marked as answer]
https://google.github.io/styleguide/cppguide.html#Integer_Types

At the C++ Standards level roughly 2/3 of committee members agree that
unsigned-sizes was a mistake, but the C++ standard is too brittle to ever
change that. Non-standard types are faced with "Follow the pattern that we
believe is wrong" or "Do a higher-friction thing that is a better long-term
outcome."

(Neither answer is great.)

On Fri, Jul 30, 2021 at 5:20 PM bobaoCrbs ***@***.***> wrote:

> Hi,
> I am just wondering why the size() of btree containers is of type "signed
> size_t". (instead of just size_t?).
> i.e. this line
> <https://github.com/abseil/abseil-cpp/blob/5de90e2673bd0c19de67c68e2fe543444dc1114a/absl/container/internal/btree.h#L241>
>
> Thanks
> Bo
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/abseil/abseil-cpp/discussions/993>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC7CRX5DXF36TQDFPIXMC53T2MJRZANCNFSM5BJHS4HA>
> .
>


-- 
*If you get an email from me outside of the 9-5 it is *not* because I'm
always on or expect an immediate response from you; it is because of work
flexibility
<http://www.inc.com/john-boitnott/how-flexible-hours-can-create-a-better-work-life-balance.html>
.  Evening and weekend emails are a sign I allocated some regular working
hours for other things (such as family, gym, friends,...).  And I encourage
you to feel free to do the same.

**Reply by bobaoCrbs** [reply]
Thanks for the explanation. In that case, should we at least make it consistent across the absl container library? The hash map containers are “using size_type = size_t" (e.g., [this line](https://github.com/abseil/abseil-cpp/blob/4bb739310c0257bedc41bfe2824c3f2860398a65/absl/container/internal/raw_hash_set.h#L663))

**Reply by mbxx** [reply]
Changing the type of btree's size from signed to unsigned would be a substantial undertaking for Google's codebase.  Every place the size is used would need to be touched to make sure that the change in return type doesn't introduce a bug.

This isn't impossible.  We made a similar but larger change for the return type of `StringPiece::size()` when we changed `absl::string_view`'s API to match that of `std::string_view`.  It is a lot of work, though.

Complicating matters, we don't have a great way to help our open source users handle this change.  It's possible to write a Clang refactoring tool to help with the migration, but users who are either unaware of the change or don't use the refactoring tool are liable to have a really bad time.

The inconsistency is definitely unfortunate, but fixing it would be a lot of work and involve a significant amount of risk for very little upside.

**Reply by bobaoCrbs** [reply]
Make sense. Thanks for the explanation.
